arangod/WasmServer/WasmCommon.cpp

```diff
@@ -1,108 +1,150 @@
-        isDeterministicField.errorNumber(),
-        functionName + std::move(isDeterministicField).errorMessage());
+        isDeterministic.errorNumber(),
+        functionName + ": Field 'isDeterministic': " +
+            std::move(isDeterministic).errorMessage());
   }
-  auto isDeterministic = isDeterministicField.get();
 
-  return ResultT<WasmFunction>(WasmFunction{name, code, isDeterministic});
+  return ResultT<WasmFunction>(
+      WasmFunction{name.get(), {codeField.get()}, isDeterministic.get()});
 }
 
-auto WasmFunction::checkOnlyValidFieldnamesAreIncluded(
-    std::set<std::string>&& validFieldnames, Slice slice) -> Result {
+auto arangodb::wasm::checkVelocypack2WasmFunctionIsPossible(Slice slice)
+    -> Result {
+  if (!slice.isObject()) {
+    return Result{TRI_ERROR_BAD_PARAMETER, "Can only parse an object"};
+  }
+
+  if (!slice.hasKey("name")) {
+    return Result{TRI_ERROR_BAD_PARAMETER, "Required field 'name' is missing"};
+  }
+  if (!slice.hasKey("code")) {
+    return Result{TRI_ERROR_BAD_PARAMETER, "Required field 'code' is missing"};
+  }
+
+  std::set<std::string> validFields = {"name", "code", "isDeterministic"};
   for (const auto& field : ObjectIterator(slice)) {
     if (auto fieldname = field.key.copyString();
-        !(validFieldnames.contains(fieldname))) {
+        !(validFields.contains(fieldname))) {
       return Result{TRI_ERROR_BAD_PARAMETER,
-                    "Found unknown field " + fieldname};
+                    "Found unknown field '" + fieldname + "'"};
     }
   }
   return Result{};
 }
 
-auto WasmFunction::requiredStringField(std::string const&& fieldname,
-                                       Slice slice) -> ResultT<Slice> {
-  if (!slice.hasKey(fieldname)) {
-    return ResultT<Slice>::error(TRI_ERROR_BAD_PARAMETER,
-                                 "Required field " + fieldname + " is missing");
+auto arangodb::wasm::velocypack2Name(Slice slice) -> ResultT<std::string> {
+  if (slice.isString()) {
+    return slice.copyString();
+  } else {
+    return ResultT<std::string>::error(TRI_ERROR_BAD_PARAMETER,
+                                       "Should be a string");
   }
-  if (auto field = slice.get(fieldname); field.isString()) {
-    return field;
+}
+
+auto arangodb::wasm::velocypack2Code(Slice slice) -> ResultT<Code> {
+  if (slice.isArray()) {
+    std::vector<uint8_t> code;
+    for (auto const& p : velocypack::ArrayIterator(slice)) {
+      if (p.isInteger() and p.getInt() >= 0 and p.getInt() < 256) {
+        code.emplace_back(p.getInt());
+      } else {
+        return ResultT<Code>::error(TRI_ERROR_BAD_PARAMETER,
+                                    "Array should include only bytes");
+      }
+    }
+    return Code{code};
+  } else if (slice.isString()) {
+    auto string = slice.copyString();
+    if (!regex_match(string, std::regex("^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/"
+                                        "]{3}=|[A-Za-z0-9+/]{2}==)?$"))) {
```

Comment:
Where does this regex come from? maybe it should be a constexpr somewhere in a header file?

Reply:
I guess the best fit is to add this constexpr to arangodb::baseics::StringUtils.
I just googled a bit, found this regex and also checked it against the description on wikipedia on how base64 works. It makes sense: A base64 string consists of letters A-Z, a-z, 0-9, + or /. At the end there can be a padding of one or two = symbols. This padding results from the fact that base64 converts 8-bit symbols into 6-bit symbols. This can be done without a rest when 3 8-bits are converted into 4 6-bits (therefore the first block in the regex checks for blocks of 4). If the full number of input bits is not a multiple of 3, the rest is dealt with via = symbols: 2 8-bits are converted into 3 6-bits and one = (the second block in the regex) and 1 8-bit is converted into 2 6-bits and two = (the third block).

Reply:
TODO: Write own decodeBase64 function that returns a vector<uint8_t>. The current StringUtils::decodeBase64 can only be used for encoded strings because: If the encoded value includes a nullbyte, the current decode function does not return the stuff after this nullbyte.

Reply:
My previous comment was not true: I added tests that verify that base64 de- and encoding works also for strings that include nullbytes.